pkg/topology/kademlia/kademlia.go

```diff
@@ -668,28 +671,28 @@ func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr,
 func (k *Kad) Announce(ctx context.Context, peer swarm.Address) error {
 	addrs := []swarm.Address{}
 
-	_ = k.connectedPeers.EachBinRev(func(connectedPeer swarm.Address, _ uint8) (bool, bool, error) {
-		if connectedPeer.Equal(peer) {
-			return false, false, nil
-		}
+	r := rand.New(securerand.NewSource())
+
+	for bin := uint8(0); bin < swarm.MaxBins; bin++ {
 
-		addrs = append(addrs, connectedPeer)
-
-		// this needs to be in a separate goroutine since a peer we are gossipping to might
-		// be slow and since this function is called with the same context from kademlia connect
-		// function, this might result in the unfortunate situation where we end up on
-		// `err := k.discovery.BroadcastPeers(ctx, peer, addrs...)` with an already expired context
-		// indicating falsely, that the peer connection has timed out.
-		k.wg.Add(1)
-		go func(connectedPeer swarm.Address) {
-			defer k.wg.Done()
-			if err := k.discovery.BroadcastPeers(context.Background(), connectedPeer, peer); err != nil {
-				k.logger.Debugf("could not gossip peer %s to peer %s: %v", peer, connectedPeer, err)
+		connectedPeers := randomSubset(r, k.connectedPeers.BinPeers(bin), broadcastBinSize)
```

Comment:
It'd be great to see the outbound connector use something like this instead of the current BinRev sorted peerID selection of outbound peer to connect to.  Especially for low-numbered, distant bins!

Reply:
For details of the connection behavior, see the following capture.  Look at the first peers in buckets 0, 1, and even 2 ignoring the few that were randomly added by gossip after the restart.  Code to generate your own connection peer list is included in:

https://ipfs.io/ipfs/QmSKgFpoXZoGJHo2DXhPoxikXdfZc8ye5KPg7AjnHiYbtA?filename=peer-buckets.txt

Reply:
From the perspectives of two nodes that are close and share a neighborhood, there could be performance benefits to randomizing the peers list because they chances are, they share a similar kademlia topology, hence a similar peer list in each bin.